## 1. Why Use ConfigBox Instead of a Dictionary?

In [1]:
import numpy

In [5]:
example = {"key1":"val1", "key2":"val2"}

In [6]:
example["key1"]

'val1'

In [8]:
example.key1

AttributeError: 'dict' object has no attribute 'key1'

#### 1. A normal dictionary only supports key-based access (example["key1"]); using dot notation (example.key1) raises an AttributeError.
#### 2. ConfigBox wraps the dictionary and enables attribute-style access (example.key1), making the code cleaner and easier to read.
#### 3. This is why we prefer ConfigBox for configuration files—it provides more convenient and readable access without changing the underlying data.

In [9]:
from box import ConfigBox
example = ConfigBox({"key1":"val1", "key2":"val2"})
example.key1

'val1'

## 2. Why use Ensure annotation?

In [10]:
def get_product(x:int, y:int) -> int:
    return x * y

In [11]:
get_product(2, 2)

4

In [ ]:
get_product(4, "2") ## Problem: Prints 4 times 2

'2222'

#### 1. Python type annotations (x: int, y: int) are only suggestions, so Python still accepts a string like "2" and returns "2222" instead of an error.
#### 2.The @ensure_annotations decorator checks whether the values passed to the function match the expected data types.
#### 3.If the wrong data type is passed, it immediately raises an error, helping you find mistakes early and write safer code.

In [14]:
from ensure import ensure_annotations
@ensure_annotations
def get_product(x:int, y:int) -> int:
    return x * y

In [15]:
get_product(4, "2") ## Problem: Prints 4 times 2

EnsureError: Argument y of type <class 'str'> to <function get_product at 0x0000027FDF569A80> does not match annotation type <class 'int'>